### ISSUE TITLE
bisync: auto-resolve conflicts

### ISSUE THREAD

**Issue by nielash** [rclone/rclone#7471]
This is a proposal for a new bisync feature to optionally auto-resolve sync conflicts (and customize bisync's change detection behavior more generally), inspired in part by discussions with @crocinsocks on #7332 and #5681.

### What problem are you are trying to solve?
Currently, when a file is new/changed on both paths (relative to the prior sync), and the versions on each side are not identical, bisync will keep both versions, renaming them with `..path1` and `..path2` suffixes, respectively. Many users have requested more control over how bisync handles such conflicts -- including an option to automatically select one version as the "winner" and rename or delete the "loser".

Similarly, users have requested more control over how conflicting files are handled during `--resync`. Currently the `path1` version prevails by default, but users have requested options to automatically select the winner based on characteristics such as `newer`, `older`, `larger`, and `smaller`.

Some users have additionally requested the ability to use the above-described auto-resolve features _as an alternative to needing `--resync` at all._ In other words, in an error scenario where bisync would otherwise demand a manual `--resync` to recover, this option would tell bisync to **automatically** run a `--resync` and auto-resolve any conflicts according to the rules described above. Such an option would maximize resilience at the possible expense of perfection (for example, [deleted files could occasionally re-appear](https://github.com/rclone/rclone/issues/5681#:~:text=Bidirectional%20sync%20can,has%20suddenly%20reappeared%3F).) Because bisync takes an approach of "safe by default", such an option would be disabled by default, but made available to advanced users via a new flag.

### How do you think rclone should be changed to solve that?

I propose the following new options:
```
--auto-resolve CHOICE     Automatically resolve conflicts by preferring the version that is: 
                            newer | older | larger | smaller | path1 | path2 (default: none)
--conflict-loser CHOICE   When --auto-resolve selects a winner, do this to the loser:
                            delete | num | pathname (default: num)
``` 
For example, suppose there is a sync conflict of `file` and the version on Path1 is newer.

- With `--auto-resolve newer`, `path2/file` is renamed to `path2/file..conflict1` (or, if that already exists, `path2/file..conflict2`, etc.) and copied to path1, and `path1/file` is copied to path2 (without renaming).
- If `--conflict-loser pathname` is added, path2 version is renamed to `path2/file..path2` instead of `path2/file..conflict1` (overwriting any `path2/file..path2` that already exists.)
- If `--conflict-loser delete` is added, path2 version is deleted instead of renamed.

Options specific to `--resync`:
```
--resync-prefer OPTION    During --resync only, prefer the version that is: newer | older | larger | smaller | path1 | path2 (default: path1)
--auto-resync             If set, automatically run --resync to recover after critical errors. WARNING: only use if you are SURE you know what you are doing! (default: false)
--max-lock DURATION       Consider lock files older than this to be "expired" (default: 1d if --auto-resync=true, otherwise 0 (never expire))
```
`--resync-prefer` and `--auto-resync` can be used together, and either one can also be used without the other.

`--max-lock` is a compromise necessary to balance `--auto-resync` with [lock files](https://rclone.org/bisync/#lock-file). Otherwise, if `--auto-resync` always respected lock files, a single interrupted run would break `--auto-resync` forever. But if `--auto-resync` always _ignored_ lock files, it could start resyncing while a prior non-resync is still running. The implementation of `--max-lock` could possibly include a mechanism for alive runs to renew themselves periodically, to determine with more certainty whether a lock file was left by an interrupted run or one that is still running, just taking awhile.

Note that the `newer` and `older` options require modtime support for both paths, and will be ignored if either remote lacks modtimes (or if the user has requested to ignore them with [`--checksum`](https://github.com/rclone/rclone/issues/5683) or [`--size-only`](https://github.com/rclone/rclone/issues/5684).) 

An open question is what should happen when a winner can't be determined by the requested method. For example: `--auto-resolve newer` where the sizes differ but modtimes are identical. Probably the easiest would be to revert to the current behavior (keep both, do `..path1` / `..path2` rename), but I am open to other ideas!

### Further thoughts

Bisync users tend to fall into one of two camps:

- User A demands perfection every time. If Bisync is ever unsure of what to do, User A would prefer that Bisync stop and ask for help rather than take its best guess and continue.
- User B wants to "set it and forget it", and values resilience above perfection. Often this user is syncing something like a movie collection which could be easily re-downloaded from the internet or backups if necessary, so total perfection isn't required, and sync conflicts may not be worth the time to resolve. User B prefers that Bisync take its best guess and continue, rather than stopping to bother them every time it has a question, or worse -- refusing to run again until it gets an answer.

Bisync in its current form is a great fit for User A, and likely to cause a bit of frustration to User B. The features discussed above are aimed at making both camps happy -- it is User A's way by default, but User B can make it behave their way by adding flags.

Fellow bisync users, feel free to comment here with any thoughts on the above! I'd love to consider any and all feedback before developing a proof-of-concept (which I hope to do shortly.)

### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

**Comment by crocinsocks** [comment]
Thank you nielash for the great write up here.

My initial thoughts on this are:

1. Flag names should be named based of relation to others as it gets very confusing otherwise so I have renamed the flags below. Subjective but believe is a valid choice.

2. In regards, to further thoughts you write: 

> User A demands perfection every time. If Bisync is ever unsure of what to do, User A would prefer that Bisync stop and ask for help rather than take its best guess and continue.

And then suggest this is how bisync currently works. However, this is not the case bisync currently renames files rather than stopping and erroring and does this with a completely random naming scenario e.g. .path1, .path2. This is fundamentally flawed as per my previous comments as .path1, .path2 is meaningless in a multi-user, multi-pc setup, the very nature of bisync and I believe should be changed entirely to conflict1, conflict2 as default. I agree that if bisync did do as you have written it would "safe by default" and would prevent sync on conflicted files and log this as an error. See prevent option added below.

3. Handling Conflicts

```
--conflict-winner CHOICE  Automatically resolve conflicts by preferring the version that is: If none both files are renamed or sync prevented as set by --conflict-loser
                            none | newer | older | larger | smaller | path1 | path2 (default: none)

--conflict-loser CHOICE   When conflict choose how to handle. When --conflict-winner none 'delete' is ignored and 'num' chosen. 
                            prevent | delete | num (default: num)
                            
--conflict-name	STRING   Choose a name for conflicts. Use comma separated string to name path1,path2 sperately e.g. 'user,cloud' (default: 'conflict')
                            
--conflict-shortname	 Will rename files using .c1, .c2, ... over .conflict1, .conflict2, ... if --conflict-loser set to num and --conflict-name unset
```

This would also mean losing conflicts get automatically renamed conflict1, conflict2 and so on by default and there is no winner. Same as now but without the issues from point 2. Whilst I think that this default is fundamentally unsafe so is any automatic conflict resolution and is purely subjective about what is best. I will put forward in this thread the default in most other commonly used bisync (OneDrive,Dropbox,GDrive,etc..) would be `newer` and `num`. Other power user tools have different mechanism such as unison to either of these choices.

I do not think there should be any overwrite of conflicts so for example when custom naming is used it would be something like ('user,cloud') user1, user2, cloud1, cloud2 and so on.. Personally, I doubt I would ever use custom naming as one side is always going to lack meaning and there is no viable way for that to be otherwise.

I think the concept of .path1 must be removed in its entirety due to points previously mentioned.

--conflict-shortname maybe this should be automated and only used for files names above x but using this flag this could be forced always. 

3. Resync

I think the resync flag should be a singular flag not two and same if auto-resync is implemented:

```
--resync OPTION    newer | older | larger | smaller | path1 | path2 (default: path1)
```

Otherwise, not much to say on this flag. I think to keep backwards compatibility 'path1' will have to remain as default. 

4. Auto-Resync

I know I initially pushed for an `--auto-resync` due to the resync being required too much however after conversations and more time I think this feature is inherently useless and dangerous and so it may be best to avoid such a design. After both filters and exits (#7470) are handled I do not really seem a use case for `--resync-auto` as this should never be required short of listing being missing and when `--resilient` flag used.

For filter changes, I would prefer something like an error stating please choose a `--resync-filters` (if not already present) option rather than succumb to a global `--resync-auto` flag and should only apply to the changes in filters rather than resyncing all filters again. 

```
--resync-auto OPTION    newer | older | larger | smaller | path1 | path2 (default: path1)
```

5. Checksum and SizeOnly

> Note that the newer and older options require modtime support for both paths, and will be ignored if either remote lacks modtimes (or if the user has requested to ignore them with --checksum or --size-only.)

I agree with this and is simply a limitation that can't be avoided.

6. Identical Sizes or Times 

> An open question is what should happen when a winner can't be determined by the requested method. For example: --auto-resolve newer where the sizes differ> but modtimes are identical. Probably the easiest would be to revert to the current behavior (keep both, do ..path1 / ..path2 rename), but I am open to other ideas!

Bar the fact I think it should be .conflict1, .conflict2 by default, I agree. Though perhaps, there could a --conflict-winner-fallback set to something like 'larger' | 'smaller' when 'newer' | 'older' for --conflict-winner and vice versa for the other way round and only after all these fail it renames both files.

6. Lock files

Lock files I think should be handled in another thread and completely separate to any of this work. Personally, when local it should work on the basis of checking if current processID is running and is the same command (same being bisync path1 path2 e.g. flags ignored) using some form of regex. This mechanism can be expanded to handle a time based mechanism when/if remote --workdir are implemented.

Thanks

**Comment by nielash** [comment]
Thank you as always for making me challenge my assumptions! 😄 I agree with a lot of what you've said here. Some responses below:

 > Flag names should be named based of relation to others as it gets very confusing otherwise so I have renamed the flags below.

Agree, and I like many of your suggestions. Allowing a customizable `--conflict-name` is a great idea (though I wonder if something like `--conflict-suffix` might be clearer? And more rclonic given [`--suffix`](https://rclone.org/docs/#suffix-suffix)?) 

I like pairing `--conflict-winner` and `--conflict-loser`, but do you think it might be confusing that one refers to a comparison method and the other refers to an action?

What would your proposed `--conflict-loser prevent` do? Skip conflicts and leave the sides un-identical at the end of the sync? We would need to carefully think through how to deal with that during `--check-sync`... (would probably need to trigger [rollback](https://github.com/rclone/rclone/commit/ef4b47a9824754824a05491c3d4c7a2ef97253d4) in perpetuity until the conflict is externally resolved, and even so, it would get wiped if there were ever a `--resync`.)

I'm not sure that `--conflict-shortname` is necessary, as you could just achieve the same result with `--conflict-name c`, no?

I like the idea of `--resync OPTION` (with `--resync` alone implying `--resync path1`), however `--resync` is currently a `bool` and this would require converting it to `string`, which could possibly break people's scripts if they've been using `--resync=true` or `--resync=false` (for some reason). I would also need to check if it breaks anything with the `-1` shorthand. But it may be worth doing even so... I agree it's best not to have 2 flags when you can have 1.

> 2\. you write:
> 
> > User A demands perfection every time. If Bisync is ever unsure of what to do, User A would prefer that Bisync stop and ask for help rather than take its best guess and continue.
> 
> And then suggest this is how bisync currently works. However, this is not the case bisync currently renames files rather than stopping and erroring and does this with a completely random naming scenario e.g. .path1, .path2.

I see what you're getting at, but I don't actually think it's a contradiction (nor is the naming random). There is no "guessing" involved in Bisync's current renaming behavior -- it refuses to pick winners and losers and instead keeps BOTH versions and renames them symmetrically, giving neither one preference over the other (in fact, this refusal to pick winners/losers is your chief complaint about it, right?). The use of the pathname means you could theoretically "undo" this rename later (not that most people probably do.) But this "Further Thoughts" section was mainly referring to the `--auto-resync` idea -- naturally, User A will think it's an abomination, while User B will think the _lack_ of it is an abomination. 😄 

> This is fundamentally flawed as per my previous comments as .path1, .path2 is meaningless in a multi-user, multi-pc setup, the very nature of bisync and I believe should be changed entirely to conflict1, conflict2 as default.

> I think the concept of .path1 must be removed in its entirety due to points previously mentioned.

I'm not totally following what you mean here... how is ".path1" unsafe but ".conflict1" is not?

I get your point about multi-user / multi-pc, but I think "meaningless" is an overstatement -- suppose you have a star topology where Path1 is always the cloud and Path2 is always the local machine... this naming convention tells you that `..path1` came from the cloud and `..path2` came from a user's machine (even if you don't know exactly which user). If instead it's `file` and `file..conflict1`, there's no way to tell which came from which (although your proposed `--conflict-name` potentially offers a way around this, which I like.) 

Anyway, regardless of what we think of `..path1` / `..path2`, it has been a feature for 2+ years now and so I am not sure we can just get rid of it, for backward compatibility reasons. I do think you make a compelling case that the default should be changed -- but I think `pathname` should at least be available as a non-default legacy option.

> This would also mean losing conflicts get automatically renamed conflict1, conflict2 and so on by default and there is no winner. Same as now but without the issues from point 2. 

Just to make sure I'm following -- you mean that with the default (equivalent to `--conflict-winner none --conflict-loser num`), instead of `file` and `file..conflict1` we get `file..conflict1` and `file..conflict2`, correct? (Essentially, _both sides_ are losers?) Also, if `file..conflict2` were to then have another conflict (probably unlikely, but we need to think about it), we'd get `file..conflict2..conflict1` and `file..conflict2..conflict2`, correct? (whereas with `--conflict-winner newer` we could eventually get `file..conflict3`, `file..conflict4`, and so on.)

Also, do you think we should lpad with some number of zeroes to make sure that `file..conflict9` alphabetically sorts before `file..conflict10`? Or don't worry about it?

And regarding the `--conflict-name` (or `--conflict-suffix`), should the `..` be part of the customizable suffix, or hard-coded? Should we try to implement [`--suffix-keep-extension`](https://rclone.org/docs/#suffix-keep-extension)?

> I know I initially pushed for an `--resync-auto` due to the resync being required too much however after conversations and more time I think this feature is inherently useless and dangerous and so it may be best to avoid such a design. After both filters and exits (https://github.com/rclone/rclone/issues/7470) are handled I do not really seem a use case for `--resync-auto` as this should never be required short of listing being missing when `--resilient` flag used.

I think I may need to give this more thought. I agree with you that `--auto-resync` is probably not something I'd ever use myself, nor do I think it's advisable generally, however for some reason it really is a request I see over and over on the forums (it's not just you)... and there are in fact some alternative sync tools that tout their lack of `--resync` as a feature, not a bug. I've even seen users code their own scripts to automatically run a `--resync` when Exit Code > 0. So I am trying to channel the "You Are Not the User" mantra and remain open-minded to the fact that some users really do seem to want this, no matter how ill-advised I may personally find it. It is hard to tell sometimes whether there's something I'm missing about their use case, or something _they're_ missing about why they do not actually want what they think they want.

Your point about [exits](https://github.com/rclone/rclone/issues/7470) is a good one, and it may be worth seeing how close to perfection we can get (particularly with recovering from ungraceful shutdowns) before making a decision about `--auto-resync`. I actually think that `--conflict-winner newer` could significantly mitigate that one edge-case I was concerned about in the "diff from last trusted listing" method. I agree that if we can recover perfectly, `--auto-resync` is probably moot. But if it's only "pretty good most of the time", I wonder if some users will still prefer `--auto-resync` due to the perceived safety of erring on the side of not-deleting (despite my total agreement with you that "no deletes" != "non-destructive").

> For filter changes, I would prefer something like an error stating please choose a --resync-filters (if not already present) option rather than succumb to a global --resync-auto flag and should only apply to the changes in filters rather than resyncing all filters again.

My current thought on this is that `--force` would be a appropriate flag than `--resync` for committing [session changes](https://github.com/rclone/rclone/issues/5678), or perhaps we need a new flag. I agree that using `--resync` as a catch-all is too heavy-handed.

FYI it is not just filters but other kinds of flags as well. For example, flags like `--ignore-case-sync`, `--size-only`, and `--drive-skip-gdocs` can all have surprising consequences if added/removed without a `--resync`.

> Lock files I think should be handled in another thread and completely separate to any of this work.

The problem is... `--auto-resync` and lock files go hand-in-hand -- we can't really implement `--auto-resync` without making some kind of change to lock files. (Same is kind of true for your [exits](https://github.com/rclone/rclone/issues/7470) ticket.) But we can table this for now until we revisit the `--auto-resync` discussion.

> Personally, when local it should work on the basis of checking if current processID is running and is the same command (same being bisync path1 path2 e.g. flags ignored) using some form of regex. This mechanism can be expanded to handle a time based mechanism when/if remote --workdir are implemented.

Not to go too far down this rabbit-hole, since we already discussed this at length 😅, but -- even if we accept your (very good) arguments about why the rest of the `--workdir` should remain local, lock files are the one thing that would benefit most from being remote, to prevent [race conditions](https://github.com/rclone/rclone/issues/7332#issuecomment-1746444947:~:text=Sync%20order%20does%20matter,machine%20gets%20there%20first.) between machines. So I'm not sure it makes sense to go to all the trouble of building PID-related logic that we already know we'll just have to remove. It's also possible that there could be issues with checking PIDs on some platforms and that the time-based dead-man's-switch option could be more robust in those cases, even on local.

**Comment by crocinsocks** [comment]
Definitely, agree with `--conflict-suffix`. Though, I would wonder whether it is actually a prefix because this is the name before a numbering but also is placed at the end of the file ergo suffix. I leave that one to you to decide. 

I agree it could be confusing perhaps it should be: 

```
--conflict-method CHOICE  Choose the method for handling conflicts 
                           renameboth | winner | prevent (default: renameboth)

--conflict-winner CHOICE  When conflict-method winner, choose how to pick the winning file.
                            newer | older | larger | smaller | path1 | path2 (default: newer)

--conflict-loser CHOICE   When conflict-method winner, choose how to handle the losing file.
                            delete | num (default: num)                     
```

I only put 'prevent' there to handle User A in your analogy. Yes, it would work in the way you state e.g. ignore these files but continue with sync though I also agree a lot of work so perhaps is a future if at all upgrade due to the complications of implementing correctly. Renaming both files is still taking an action not "do nothing and wait for user" and therefore is no different to keeping one as a "winner" and the other numbered as the user may or may not see either of these actions; though I understand the argument in very limited circumstances where dealing with an application/project folder this could cause a breakage which would as a by-product inform the user (if the application logs this correctly). Same as the argument that bisync doesn't pick a winner so is doing nothing which is correct if it then didn't take an action based on that outcome though it does by renaming both so still not userA. 

With regards to the default ".path1, .path2" I agree the naming is not random, my mistake, rather that .path1, .path2 is  seemingly meaningless. The problem with your analogy is you still do not know who made the cloud file in the first place nor the user and requires a very rigid setup procedure to potentially gain some meaning. Though this is still debatable as the only argument is not really for the cloud side (?), I can only see the users/machine side e.g. `--conflict-suffix userA` so at least you would know who or which machine to help resolve the conflict. Having a file renamed .path2 (or .cloud1) still seems to give no advantage/insight for fixing the conflict in a bisync environment as you can't determine the source. Though, the option would be there to name each path separately e.g. `--conflict-suffix userA,cloud` if one so choses.

There is also that it would not work with `--conflict-suffix path1, path2` (as a default) as you would in essence get path11, path21 or if you handle suffix ending with numbers path1-1, path2-1. I don't think due to the reasons above the old code for conflict resolution should be kept as I can't see anyone complaining about the old being removed when they can just add there own conflict suffix for both path1 and path2. The comment of being safe was specifically about 'prevent' or for unsafe about any automatic action taken whether both are renamed or only one due to a winner, not about naming: .path1 is not any safer than .conflict1, sorry if there was any confusion about this.

I started with `--conflict-shortname` before I thought of separating the flags. Reason I kept is some may just want a flag to enable .c1, .c2 over specifically setting `--conflict-suffix c` though in essence all this flag would be is a shortcut to this. The line of thinking of using a shortname in the first place is that I have file names that only just fit on the remote so adding .conflictX would cause issues and wanted to put that out there incase this is something that should automated when the name is too long? Personally, I have no issues with using `--conflict-suffix c` for a remote I know have long names and therefore is potentially unnecessary.

Great, I think with `--resync` would it not be possible for backwards compatibility to just have true = 'path1' and false still disables a resync so that all bases are covered? I think as these are params they are interpreted as strings regardless of the equal sign but I may be wrong? ...

> Just to make sure I'm following -- you mean that with the default (equivalent to --conflict-winner none --conflict-loser num), instead of file and file..conflict1 we get file..conflict1 and file..conflict2, correct? (Essentially, both sides are losers?) Also, if file..conflict2 were to then have another conflict (probably unlikely, but we need to think about it), we'd get file..conflict2..conflict1 and file..conflict2..conflict2, correct? (whereas with --conflict-winner newer we could eventually get file..conflict3, file..conflict4, and so on.)

Yes, exactly concept wise just with a change of flags if you agree with those I made at the start.. Though for example in a multi-user environment lets say you already have  file..conflict1/ file..conflict2 and file is resync 'd so you have all three and there is now another conflict I don't think it should overwrite but instead give  file..conflict3, file..conflict4 (this is when using 'renameboth')

> Also, do you think we should lpad with some number of zeroes to make sure that file..conflict9 alphabetically sorts before file..conflict10? Or don't worry about it?

I personally would not worry.

> And regarding the --conflict-name (or --conflict-suffix), should the .. be part of the customizable suffix, or hard-coded? Should we try to implement --suffix-keep-extension?

Think .. should be hardcoded and no I don't think so for `--suffix-keep-extension` though I guess thats subjective just I don't think I've seen that else where (?) and arguably one should resolve the conflict before using the file. 

Though, do we really need ..conflict1 why not just .conflict1 (/.conflict1.conflict1) ? Is there an advantage I am missing?

I agree with auto-resync, filter+other resyncs, lockfiles and think this should be separated into a new thread/issue and discussed there, perhaps three threads. Especially as once exits and the above is handled views my change on this. At least in my head there is still much to discuss on these and would also allow the above/rest (resync & conflicts) to be implemented sooner rather than seeing this as an all-in-one upgrade/issue. One thing with lockfiles on the remote is it would seem not really practical or viable in a many user setup as users would not be able to sync there files (!/?) as it would be pure luck/chance who gets there turn to sync.. I get there are certain race condition though these are limited, expected of any bisync and can't be worked around as these are more technical limitations; I would argue this is more how a user should use such a tool rather than locking the remote.

**Comment by nielash** [comment]
> I don't think due to the reasons above the old code for conflict resolution should be kept as I can't see anyone complaining about the old being removed when they can just add there own conflict suffix for both path1 and path2.

So, I think you have convinced me that the _default_ should be changed to `--conflict-loser num`, and the remaining question is only whether `--conflict-loser pathname` should stick around as a non-default option or be removed. I think that the thing I have not done a good job articulating is that `file..path1` / `file..path2` is _reversible_ in a way that `file` / `file.conflict1` is not, at least in a single-user environment (and bisync has plenty of single-users.) For example, suppose you want to restore path1 to the way it was before bisync renamed things -- you could write a simple script to find all `*..path2` files and delete, then find all `*..path1` files and rename to trim the `..path1` suffix. This is possible with `pathname` but not with `num`, as you don't know which side it came from. So, it is possible that there are some users who rely on this naming convention for this reason, and I could imagine them being unhappy if it's taken away. I think keeping it as a non-default legacy option makes everyone happy, and does not in any way disadvantage the users who choose not to use it. There are tons of rclone flags that are irrelevant to me -- I just ignore them and use the ones I want 🙂 

> Definitely, agree with --conflict-suffix. Though, I would wonder whether it is actually a prefix because this is the name before a numbering but also is placed at the end of the file ergo suffix. I leave that one to you to decide.

The same thought occurred to me, but I can't think of anything clearer. So I'm leaning towards `--conflict-suffix` (and clarify any confusion in the docs.) (After all, [`--suffix-keep-extension`](https://rclone.org/docs/#suffix-keep-extension) is not strictly a "suffix" either.)

I am not sure that `--conflict-method` helps us, as `--conflict-winner` and `--conflict-loser` would still refer to different things. It just makes 3 flags for what could have been achieved with 2. 

So, I think it is between this:
```
--conflict-winner CHOICE
--conflict-loser CHOICE
--conflict-suffix STRING[,STRING]
```

or maybe this:
```
--auto-resolve CHOICE
--conflict-loser CHOICE
--conflict-suffix STRING[,STRING]
```

I think `prevent` should be disregarded for now, but worth revisiting in the future if users are asking for it. (And if so, I think `skip` might be a clearer name for it.)

> There is also that it would not work with --conflict-suffix path1, path2 (as a default) as you would in essence get path11, path21 or if you handle suffix ending with numbers path1-1, path2-1.

I was thinking that if you wanted to keep the old behavior, you would do `--conflict-suffix path --conflict-loser pathname`. So `file` becomes:

```
file..path1
file..path2
```

And then if `file..path1` itself has another conflict (unlikely), you get:

```
file..path1..path1
file..path1..path2
file..path2
```

which is the same as what happens now. (Remember that if you want to keep the current behavior, there is no "winner" -- instead it's essentially two "losers".)

You could do the same with a different suffix, like `--conflict-suffix side`:

```
file..side1..side1
file..side1..side2
file..side2
```

and probably we would automatically omit the digits when two strings are provided, like `--conflict-suffix cloud,user`:

```
file..cloud..cloud
file..cloud..user
file..user
```
Which makes `--conflict-suffix path` behave exactly the same as `--conflict-suffix path1,path2`.

Meanwhile for your desired behavior, you would do `--conflict-winner newer --conflict-suffix c` to get:

```
file
file..c1
file..c2
```

> The line of thinking of using a shortname in the first place is that I have file names that only just fit on the remote so adding .conflictX would cause issues

This is a good point, and honestly it's another reason that users could be upset about being forced to change from `path` to `conflict`...

> incase this is something that should automated when the name is too long?

I worry that automating it could be problematic for users whose conflict-resolution workflows depend on having one consistent string they can search for. In my example above, it would have been harder to find and correct all the `..path1`s if half of them are `.p1`. I think the best solution is to let the user pick the `--conflict-suffix`, and they can choose a short one if filename length is an issue.

> no I don't think so for --suffix-keep-extension though I guess thats subjective just I don't think I've seen that else where (?) and arguably one should resolve the conflict before using the file.

I think the idea is that many systems determine the MIME Type by looking at the extension, so `file.conflict1.jpg` will be recognized as an image while `file.jpg.conflict1` will be an unreadable text file. (Resolving conflicts would probably be easier if you could preview the content...) I may look into whether it can be supported without too much trouble.

> Though, do we really need ..conflict1 why not just .conflict1 (/.conflict1.conflict1) ? Is there an advantage I am missing?

To the double-dot instead of a single-dot? No advantage that I'm aware of, other than that bisync has always used the double-dot and it's possible that it could break people's scripts if we change it. Perhaps the best compromise is to hard-code just one dot (or none?) and let people do `--conflict-suffix .path` if they want the double.

> I agree with auto-resync, filter+other resyncs, lockfiles and think this should be separated into a new thread/issue and discussed there, perhaps three threads. Especially as once exits and the above is handled views my change on this. At least in my head there is still much to discuss on these and would also allow the above/rest (resync & conflicts) to be implemented sooner rather than seeing this as an all-in-one upgrade/issue.

Happy to discuss on separate threads, but when features are closely related (as these ones are), I do think it can be helpful to think a few steps ahead before diving in -- so that feature1 can be built the right way from the start, and we don't have to totally redo it when feature2 comes along. Exits depends on lock files and auto-resync depends on both. So I think once you've had a chance to try my proposed lock files change on #7470 let's talk about it there, and #5678 is the place to discuss decoupling filter changes from `--resync`. Feel free to create a new issue for `--auto-resync` if you wish, or I'm also fine continuing to discuss it here. (FWIW it's one of the easiest changes from a technical perspective -- maybe even just 4 lines of code. The bigger question is whether it's advisable.)

> One thing with lockfiles on the remote is it would seem not really practical or viable in a many user setup as users would not be able to sync there files (!/?) as it would be pure luck/chance who gets there turn to sync.. I get there are certain race condition though these are limited, expected of any bisync and can't be worked around as these are more technical limitations; I would argue this is more how a user should use such a tool rather than locking the remote.

Firstly just to be clear, there are two different proposed changes to lock files: 1.) allowing them to expire, and 2.) allowing them to live on non-local remotes. You are focusing on issue 2, but really it's issue 1 that is the blocking issue for recovery and `--auto-resync`. 

Secondly, even if proposed change 2 were adopted, no one would be forced to use it. Each user could choose to use their own local lock file, exactly as they do today. 

Thirdly, I think there are some possible solutions to what you're describing, like perhaps some sort of central ledger to ensure a round-robin rotation. But even if there's not a good solution, I don't think that's a reason to just not worry about race conditions. There are always trade-offs, and different users have different priorities. You may prioritize more frequent syncs even if it means more race conditions, but other users may prioritize avoiding race conditions, even if it means less frequent syncs. Very much the User A / User B divide again. User B will see conflicts as just the inevitable cost of concurrent syncing, and User A will see mutual exclusion as just the inevitable cost of data integrity. They are both right 🙂  And I think bisync can make them both happy by allowing each user to choose which option is right for them.

**Comment by crocinsocks** [comment]
I think the use case described here with ..path1, ..path2 and scripts seems like a workaround for the options `--conflict-winner path1(/path2)` and `--conflict-loser delete` as they are currently not available. I think where I get to with this is it simply a paradox to say that path1 means x and path2 means cloud even in a rigid environment because which machine is path1 and which machine did path2 originally belong to? Even in a single user environment we are still talking multiple machines otherwise `copy`/`sync` should be used. I can't see any escape from this paradox and so believe it to be unsolvable. However, giving each machine/user a name in a machine(s)<->cloud can provide considerable meaning/benefit.

I agree to match suffix used elsewhere `--conflict-suffix` is probably best. 

What about `--conflict-resolve` instead of `--conflict-winner` - I think I like this and also use this later? Then perhaps `--conflict-resolve-fallback` or `--conflict-resolve-secondary`

Agree `prevent` should be ignored for now and also agree `skip` makes much more sense.

I still don't really understand the use of `pathname` in `--conflict-loser`? I can't see why anyone would want only one conflict for a file as you could have the following, due to the original file being added back without a manual resolution: (assume `--conflict-winner none --conflict-loser num --conflict-suffix conflict`)

```
file
file.conflict1
file.conflict2
```
  
Now lets say there is a conflict of the file again in the legacy way this would overwrite the already existing conflicts (dangerous and you would lose work!) where it really should end up with the following: (where c1, c2 are from the first conflict and c3,c4 are from the second) 

```
file.conflict1
file.conflict2
file.conflict3
file.conflict4
```

Whether you choose to keep a legacy flag for this behaviour is entirely up to you as you are ultimately developing the tool - personally, I would force users away from this paradox and dangerous behaviour even though I know that might cause some upset in the short run. 

I also think if you do this it should be done via `--conflict-resolve legacy` and will ultimately ignore the rest of these flags as would be better than adding `--conflict-loser pathname` as this is already handled with suffix and would try to avoid any overwriting actions in this new conflict code.

No issue with automating a shortname I agree with your points here and allows the user to be explicit in this case.

I am also certainly not against supporting `--suffix-keep-extension` if it can be without too much trouble. The idea definitely has some merit and some may like this - I may even use it though perhaps more pressing items as is more a nice to have.

I think `..` is simply another seemed good at the time but really should have never been implemented. Personally, I do not think `--conflict-suffix` should handle this and rather `..` should only be used with `--conflict-resolve legacy` (if kept) otherwise all new logic should use `.`. If you do not like this perhaps a flag the enables or disables the `..`.

Whilst I agree that recovery and auto resync rely on lockfiles its implementation should have no effect on how these are coded as it should simply call the same isLocked function. I think others may also be more likely to get involved with lock files functionality if it was on its own thread and think it is very much worthy of one. If you do not agree will layout some thoughts re lockfiles on #7470.

**Comment by nielash** [comment]
> I still don't really understand the use of `pathname` in `--conflict-loser`? I can't see why anyone would want only one conflict for a file

This suggests to me that I still have not succeeded in articulating this concept:

> `file..path1` / `file..path2` is reversible in a way that `file` / `file.conflict1` is not

and until we're on the same page about that, I fear we have a bit of a "Chesterton's Fence" scenario, where the assertion that the feature should be removed comes without a full grasp of why it is there.

> Even in a single user environment we are still talking multiple machines otherwise `copy`/`sync` should be used.

Why? There are many users who use Bisync to synchronize two local paths, or one machine with one cloud storage service (where changes can be made directly on either side -- such as via the browser UI.) `copy`/`sync` does not cover those scenarios. 

> I think the use case described here with ..path1, ..path2 and scripts seems like a workaround for the options `--conflict-winner path1(/path2)` and `--conflict-loser delete` as they are currently not available.

I don't think it is. There are some users who will want to manually review each sync conflict and decide how to resolve it. Think of it like a git merge conflict -- sometimes it's as simple as keeping one version and discarding the other, but sometimes the fix is more complicated -- like keeping a combination of both. The "User A" types will take issue with the "automatic" part, as the winners may not always follow one consistent pattern, like being from the same side, and even when they are, it can't be predicted in advance which side that will be. I think you are assuming that once these auto-resolve features are available, everyone will want to use them -- and I don't think that is true. My guess is most will, but a significant minority will prefer to resolve any conflicts manually. 

> What about `--conflict-resolve` instead of `--conflict-winner` - I think I like this and also use this later? Then perhaps `--conflict-resolve-fallback` or `--conflict-resolve-secondary`

I like `--conflict-resolve`. I think the others are too long, and `--conflict-loser` is fine if there is no `--conflict-winner` to confuse it with.

> Now lets say there is a conflict of the file again in the legacy way this would overwrite the already existing conflicts (dangerous and you would lose work!)

I see what you're getting at, but I think it assumes a false inference, as  once `file` is renamed `file..path1` and `file..path2`, there is no more `file`. That is -- a second conflict can only occur if the user has chosen not to resolve the first conflict. The overwrite behavior is consistent with how rclone's [`--backup-dir`](https://rclone.org/docs/#backup-dir-dir) works:

```
If there is a file with the same path (after the suffix has been added) in DIR, then it will be overwritten.
```

which is why many users choose to include a dated suffix like `--suffix -2019-01-01`.

Remember that in "User A"'s philosophy, a conflict is not supposed to happen -- and when it does happen, the conflicts must be resolved -- not left there forever while more conflicts are created. (I also fully understand why "User B" would prefer the convenience of letting `file.conflict48`, `file.conflict49`, etc. build up to be dealt with later or possibly never. They are both valid preferences -- just fundamentally incompatible with each other.)

> where it really should end up with the following: (where c1, c2 are from the first conflict and c3,c4 are from the second)
> 
> ```
> file.conflict1
> file.conflict2
> file.conflict3
> file.conflict4
> ```

What I think this overlooks is that there is no way to tell in this scenario which side `file.conflict3` came from, so it is not _reversible_, for the users who care about that. Which is not me saying this option shouldn't exist -- it should! But I disagree that it should be the _only_ option.

> personally, I would force users away from this paradox and dangerous behaviour even though I know that might cause some upset in the short run.
> 
> I also think if you do this it should be done via `--conflict-resolve legacy` and will ultimately ignore the rest of these flags

Naming it "legacy" says to me "if we were starting over today, we would just remove this", and I don't think that's actually what this is. I think it's a valid option that serves an unfilled need for a minority of users -- but just shouldn't be the default.

The thing to keep remembering is that two users with different preferences will often both think their way is "right" and that all other users should be forced to use it. The maintainer's job is to balance the needs of all users, which sometimes conflict (no pun intended). Often that means making both options available. This becomes untenable if users start saying "it's not enough for me to get my option, I also need the other guy's option removed." As I said, there are a number of rclone flags that make me think "I can't imagine why anyone would possibly want that", but I ignore them and move on, knowing that some users probably feel the same way about flags I find essential for some unusual reason 🙂 

> would be better than adding `--conflict-loser pathname` as this is already handled with suffix

I think what you're saying is that `--conflict-loser pathname` is unnecessary because it is already implied by `--conflict-suffix path1,path2`, right? I think the issue with this is that if there is no `pathname` then users are forced to use `--conflict-loser num`,  so if I want (say) `--conflict-resolve newer --conflict-suffix path1,path2` then I end up with:

```
file
file.path11
file.path12
file.path21
file.path22
file.path23
```

with no way to disable the auto-numbering. So essentially, `--conflict-loser pathname` is a way to say "do not auto-number, keep only the most recent conflict". But I agree this probably wouldn't come up too much, as most users who want `pathname` would probably also want `--conflict-winner none`. And others might choose to add a hyphen/underscore, like `--conflict--suffix path1_,path2_`



> I think `..` is simply another seemed good at the time but really should have never been implemented.

I tend to agree, although I think rather than adding another flag, the most efficient way to handle this is to say it's one dot by default, and if you want more dots you can simply add them into your `--conflict-suffix`. I don't see how that in any way harms users who choose to stick with the default.

> Whilst I agree that recovery and auto resync rely on lockfiles its implementation should have no effect on how these are coded as it should simply call the same isLocked function. I think others may also be more likely to get involved with lock files functionality if it was on its own thread and think it is very much worthy of one. If you do not agree will layout some thoughts re lockfiles on https://github.com/rclone/rclone/issues/7470.

I'm not sure if I follow what you mean about the function. Which thread we use is not terribly important to me (feel free to start a new one if you prefer), but I do think that a prerequisite for further lock files discussion should be you trying out the `--max-lock` implementation I put forward on #7470, which I think is quite good.

**Comment by crocinsocks** [comment]
I think some of this has been taken out of context for example you were specifically talking about a user not manually resolving but using a script to automatically rename and the method I suggested would do so in the way I believe you intend rather than the method you suggest that would have the issue as in No.2 below.

For clarity any use of the word user,machine also implies client whether that be separate to rclone e.g. a web ui or any other means and would imply bisync. So user/client/machine all have the same meaning.

Would seem we agree on and for ease am going to use shortened versions as below:

```
--conflict-resolve (-cr)
--conflict-loser (-cl)
--conflict-suffix (-cs)
```

I think it may be better to ask do you agree that ... questions and then go from there?

--

So do you agree that:

1) It is paradoxical to assign any meaning to any side of conflict resolution in a bi-synchronous environment bar specific and unique client ids

2) With `-cr none -cl num -cs conflict` there is no difference in meaning `-cr none -cl num -cs path` (unless the code randomly assign the number)

```
file.conflict1 = file.path1 (path1)
file.conflict2 = file.path2 (path2) 
file.conflict3 = file.path3 (path1) 
file.conflict4 = file.path4 (path2)
// or with nested conflicts
file.conflict1.conflict1 = file.path1.path1 (path1)
file.conflict1.conflict2 = file.path1.path2 (path2)
```

And therefore both are reversible in the method you describe (though the method you describe may not have the result I believe you think it does due primarily to point No.1 and No.3)

3) That in the legacy conflict mechanism it is more than possible for both `file` and `file..path1` and `file..path2` to exist and there for it will be destructive if another conflict of `file` happens

For example, having `file` open in a program/editor, a conflict happening and then saving which would write back to `file`. Or in a multi-user environment editing a file before the conflict has been updated. An application that auto recreates a file when missing. I'm sure there are many others. Not that it can only occur if a user has decided not to handle the file.

4) That there are really three types of user for conflict resolution: 

```
A - Do nothing and ask user 
B - Handle conflict and always keep losers (safe)
C - Handle conflict and delete or overwrite losers (destructive)
```
Handle conflict means the concept of winner/losers whether both lose (e.g. renaming both) or one wins (e.g. either way an action is taken)

5) That only the sync exiting with an error code and listing the errors with something like '-cr skip' (or interactive prompt `-cr ask`) is the only way to be UserA?

6) That `-cr none -cl num -cs conflict` & `-cr newer -cl num -cs conflict` are both UserB?

7) That the legacy way of handling conflicts is inherently UserC?

8) That you could get *similar* behaviour to legacy by setting a flag such as `-cr none -cl overwrite -cs path (or .path for ..)`

Overwrite would only allow a singular conflict to exist only ergo legacy so you would get `file.path1` and `file.path2` regardless of the amount of `file` conflicts and with any other `-cr` where one wins `file` & `file.path1` (never `file.path2`, `file.path3`, ...). 

9) That just because `-cr none -cl num -cs conflict` renames both files that does not imply that the user has been informed of this action. 

Though due to lack of access of this file through certain means may in some cases my make it more noticeable though could also cause breakages that go unnoticed too. If looking for the file via an explorer any resolution where the loser is kept would be obvious.

10) That because you can't gain any meaning with .path1/.path2 (if you agree with the relevant points above) that doing so creates a false sense of security and could mislead users? 

-- 

Hopefully, answering yes/no to the above will give clarity where we disagree, if we do and help hash out anything left remaining. 

-- 

Will start a thread re lockfiles so not to conflate - thanks and will explain the bit about the isLocked function.. Whilst I haven't test your solution I have looked over the code and do understand how it functions :)

**Comment by nielash** [comment]
> you were specifically talking about a user not manually resolving but using a script to automatically rename and the method I suggested would do so in the way I believe you intend

The script example was about _reversing_, which is different from _resolving_. _Reversing_ means undoing whatever mess bisync may have made, with the goal of reverting both paths back to how they were the moment before bisync ran. This still leaves you with conflicts, which is why it's not _resolving_. But it matters from User A's perspective because an irreversible change is inherently more destructive than a reversible one.

> 1. It is paradoxical to assign any meaning to any side of conflict resolution in a bi-synchronous environment bar specific and unique client ids

It's possible I'm misunderstanding what you mean, but no -- I don't think it's paradoxical. Suppose I'm a single user with one bisync run, where Path1 is Google Drive and Path2 is my local computer. Path1 and Path2 have clear meanings to me in this case. When I see `file..path1`, I know it was the Google Drive side of the conflict. When I see `file..path2`, I know it was the local computer side of the conflict. It's helpful info for me to have when deciding which version to keep.

> 2\. With `-cr none -cl num -cs conflict` there is no difference in meaning `-cr none -cl num -cs path` (unless the code randomly assign the number)

I agree that whether the suffix is "path" or "conflict" doesn't make much difference. But I disagree that an auto-numbering system is as reversible as the legacy behavior. I think your point is that we can deduce that 3 must be Path1 because it is odd, and 4 must be Path2 because it is even? That makes a lot of assumptions -- like that the files were always renamed in the same order, and that none were ever deleted. All it would take is one missing file somewhere to break the pattern for all subsequent files.

> 3. That in the legacy conflict mechanism it is more than possible for both file and file..path1 and file..path2 to exist and there for it will be destructive if another conflict of file happens

Agree that it's possible, disagree that it's a major concern. The scenario you describe requires another process to intervene and create a secondary conflict. Users can of course have all kinds of processes on their machines that cause chaos to their files -- at a certain point, that's on them, and not something we can control. And I'm not sure how your solution improves that -- doesn't `file` still get overwritten with `-cr newer` (in your scenario where it is open during the sync and re-saved after)? If the behavior of overwriting a prior suffixed conflict is destructive, then by definition, do you also think [`--backup-dir`](https://rclone.org/docs/#backup-dir-dir) is destructive?

> 4\. That there are really three types of user for conflict resolution:
> 
> ```
> A - Do nothing and ask user 
> B - Handle conflict and always keep losers (safe)
> C - Handle conflict and delete or overwrite losers (destructive)
> ```

Mostly agree, but I think the "safe" and "destructive" characterizations are too simplistic. Does B keep the losers in a reversible or irreversible way? Does C delete the losers always, or only if the user has allowed a secondary conflict to occur, without resolving the first conflict? 

Here is how I tend to think of it instead:
```
A - Do not pick winners and losers -- either stop and ask user or take some action that user could undo later (like move files out of the way *reversibly*)
B - Use best efforts to pick a winner, and either delete or irreversibly rename the loser
```

I think part of the reason we may be talking past each other is because you are treating "keep" and "delete" as synonymous with "safe" and "destructive". Suppose that some malware program randomly renames every file on your computer without changing the contents -- do you consider this "safe" simply because no files have been deleted?

Now, consider another program that does nearly the same thing, except it's a filename encryption program and you hold the decryption key. The reversibility makes a difference, no? This is why I am more focused on reversible vs. irreversible -- not all "changes" are bad, and not all "keeps" are good.

What bisync's legacy behavior aims to do is move the conflicts out of the way in a way that the user can later revert, should they choose. Akin to encrypting a filename and giving the user a decryption key (which yes, is still a "change", but a harmless one).

> 5\. That only the sync exiting with an error code and listing the errors with something like '-cr skip' (or interactive prompt `-cr ask`) is the only way to be UserA?

Partially agree. I think there probably are some User A types who would see this as the only acceptable solution. But other User A types who would find the legacy behavior acceptable. Not all User As are exactly the same. I would note that in the original Bisync docs, this is how it is described:

> ### Safety measures
- Lock file prevents multiple simultaneous runs when taking a while. This can be particularly useful if bisync is run by cron scheduler.
- Handle change conflicts non-destructively by creating `..path1` and `..path2` file versions.

> 6. That `-cr none -cl num -cs conflict` & `-cr newer -cl num -cs conflict` are both UserB?

Agree.

> 7\. That the legacy way of handling conflicts is inherently UserC?

Disagree, because of the way it's worded. This more accurately describes `-cr newer -cl delete`. The legacy way does not "handle" conflicts (it more _refuses_ to handle them), it does not delete, and it is not inherently destructive. I do understand why you see it that way -- because you are focusing on a possible scenario where a secondary conflict overwrites an unresolved prior conflict, and I agree that is theoretically possible, but I guess I see that more as preventable user error that is out of scope for bisync to worry about. What I will say is that if you have suggestions for how to better handle this scenario in bisync, I am all ears and happy to consider improvements. 

> 8\. That you could get _similar_ behaviour to legacy by setting a flag such as `-cr none -cl overwrite -cs path (or .path for ..)`

Agree, given the way you've defined "overwrite", although I don't think "overwrite" is an accurate label for the legacy behavior. Again, nothing is overwritten under normal circumstances. The scenario you're focusing on requires a secondary conflict to occur before the original conflict is resolved. It is "overwrite" only in the same sense that `--backup-dir` is "overwrite". Is it perfect and totally robust to all kinds of edge cases? No. But I do think it is pretty good, and better (for this particular goal) than any alternatives I've seen suggested in this thread. 

> 9. That just because `-cr none -cl num -cs conflict` renames both files that does not imply that the user has been informed of this action.

Agree. But it would be quite easy to identify them with, for example:
```
rclone lsl remotename --include "*.conflict*"
```

Personally I have a script that runs weekly and emails me a report of such issues, and I suspect that more serious User A types run it more frequently than that.

> 10. That because you can't gain any meaning with .path1/.path2 (if you agree with the relevant points above) that doing so creates a false sense of security and could mislead users?

Disagree with the premise -- as I said in number 1, I think it's quite possible to gain meaning from `..path1`/`..path2`. When you say false sense of security, do you mean because of the secondary conflict overwrite scenario? If so, perhaps adding a note to the docs would help, similar to the one `--backup-dir` has? I'm a little confused because on the one hand you seem concerned that the legacy behavior isn't safe _enough_, but on the other hand you are saying it is _too_ safe, to the point of being useless, and you are advocating removing it and forcing all users to use a far less cautious / reversible method. 

Here are my two questions for you in return:

1. Let's assume for the purposes of these questions that we agree that the legacy behavior is utterly pointless, very likely destructive, should never have been released, and anyone still wanting to use such an option must not be very smart. Do you agree that it does not harm you personally if it continues to exist as a non-default flag that you will never be forced to use?

2. Do you agree that it is ever acceptable to develop/maintain an optional feature because other users want it, even if you personally find their use of it to be misguided or ill-advised?

**Comment by crocinsocks** [comment]
I do not have much time, its been hectic on the run up to Christmas (as always) though I'd like to respond quickly.

This is 100% resolving not reverting, the only way to revert is to cache previous changes including deletes and restore both local and cloud changes as well as the listings. I think you mean resolve conflicts as local so in essence you would have "partially-reverted" only the conflicts back to previous state.

> 1) It is paradoxical to assign any meaning to any side of conflict resolution in a bi-synchronous environment bar specific and unique client ids

What you just described is (assumed) unique naming due to rcloneClients=1 so seems we agree.

> 2) With -cr none -cl num -cs conflict there is no difference in meaning -cr none -cl num -cs path (unless the code randomly assign the number)

I agree they both lack meaning which is kind of the point. You keep talking about reversing but there is currently no way in rclone to correctly reverse only resolve. The point is per no 1. the only way to resolve in this way is to have unique names and use a more destructive `-cr none -cl overwrite`. You are correct that the numbering is an issue so it may be worth another options that always number path1 as odd and path2 as even but again I'm still unsure of what benefit this would provide.

> 3) That in the legacy conflict mechanism it is more than possible for both file and file..path1 and file..path2 to exist and there for it will be destructive if another conflict of file happens

It is actually quite a major concern as this will happen in almost any program people regularly use from Libre/MS Office, IDEs,DAWs,Video Editors, etc... Re `-cr newer` depends on the program; most IDEs will alert you when the file has changed on disc. Though, this point is relevant to the legacy way where in this case files would be lost though with `-cr none -cl num` all future conflicts would still be kept. I know nothing regarding --backup-dir and I'm not sure it has any relevance here we aren't talking about backups but a live bi-synchronous environment. You'd be surprised but if you have your sync running regularly and switch between machine regularly this is likely one of the most common 'edge' case one might experience. Or what about somebody running a resync that has not yet had the conflict. This can and does happen. Both these events have happened to me in my short time of using rclone. 

> 4) That there are really three types of user for conflict resolution:.. 

> Does C delete the losers always, or only if the user has allowed a secondary conflict to occur, without resolving the first conflict?

Doesn't matter it has the potential to overwrite so whilst first conflict may be userB it belongs in userC due to this fact.

> I think part of the reason we may be talking past each other is because you are treating "keep" and "delete" as synonymous with "safe" and "destructive". Suppose that some malware program randomly renames every file on your computer without changing the contents -- do you consider this "safe" simply because no files have been deleted?

Now this is an irrelevant case (unlike No.3) and very much outside conflict resolution or rclone scope and the reason one needs archives / backups. I still support my use case of keeping files as safe and automatically deleting or create with overwrite as destructive.

> Now, consider another program that does nearly the same thing, except it's a filename encryption program and you hold the decryption key. The reversibility makes a difference, no? This is why I am more focused on reversible vs. irreversible -- not all "changes" are bad, and not all "keeps" are good.

Again, you are not reversing but resolving which may give the impression of a partial revert.

> 5) That only the sync exiting with an error code and listing the errors with something like '-cr skip' (or interactive prompt -cr ask) is the only way to be UserA?

> Partially agree. I think there probably are some User A types who would see this as the only acceptable solution. But other User A types who would find the legacy behavior acceptable. Not all User As are exactly the same. I would note that in the original Bisync docs, this is how it is described:

I think there are no userA that would agree with legacy behaviour (userC) but would accept any of userB.

The docs are wrong the current method has the potential to be destructive it is by very nature overwrite.

> 7) That the legacy way of handling conflicts is inherently UserC?

Don't want to sound like a broken record and already answered my reasoning for this statement. 

> 8) That you could get similar behaviour to legacy by setting a flag such as -cr none -cl overwrite -cs path (or .path for ..)

Overwrite in this case means to create or update I was thinking `upsert` but that is never really used outside of SQL. In most programs similar / file based overwrite works on the basis create a file if not exists if it does allow update/overwrite and without overwrite it errors and stop/skips if file currently exists. So I would say current behaviour is overwrite. Unless you have a better name?

> 9) That just because -cr none -cl num -cs conflict renames both files that does not imply that the user has been informed of this action.

> Agree. But it would be quite easy to identify them with, for example: 

You would still have this unless using `-cl delete` e.g. no greater insight if one file or both are renamed. 

> 10) That because you can't gain any meaning with .path1/.path2 (if you agree with the relevant points above) that doing so creates a false sense of security and could mislead users?

> Disagree with the premise -- as I said in number 1, I think it's quite possible to gain meaning from ..path1/..path2. 

Already, shown that your example in No. 1 ( rcloneClients=1) implies these names would be unique and there provide some meaning. 

> When you say false sense of security, do you mean because of the secondary conflict overwrite scenario? 

No I mean that by default when rcloneClients=n (n>2) these have no meaning only unique client names do so it is a false sense of security.

> I'm a little confused because on the one hand you seem concerned that the legacy behavior isn't safe enough, but on the other hand you are saying it is too safe, to the point of being useless, and you are advocating removing it and forcing all users to use a far less cautious / reversible method.

Not at all I am saying that the legacy method is UserC because it has the potential to overwrite and there is very little difference to any UserB options so whatever the default is, is just preference. I've also said if someone wants the overwrite method of course that's fine and am yet to see any reasoning that `-cl overwrite` would not provide the `legacy` behaviour. 

Your questions:

> 1. Let's assume for the purposes of these questions that we agree that the legacy behavior is utterly pointless, very likely destructive, should never have been released, and anyone still wanting to use such an option must not be very smart. Do you agree that it does not harm you personally if it continues to exist as a non-default flag that you will never be forced to use?

It not about being smart or whether it is pointless. Some people may prefer only having one (or two when both lose) conflict at a time I assume this is only for those with regular archives who aren't scared of losing conflicts. The issue is the fact that what is clearly UserC and destructive by nature is being referred to as safe and UserB. Of course, to also provide legacy compatibility with the upgrade users should have the option of `cl overwrite` though should be informed of the risks and also that without unique naming you can not resolve a conflict in the manner suggested e.g automated via script nor can you ever revert though you can consider this type of resolving as perhaps a partial revert although personally I think that is still wrong to use the word revert at all. Right now there is only one way to revert which would be restore previous state from an archive/backup and run a sync (prehaps a --resync ...). 

> 2. Do you agree that it is ever acceptable to develop/maintain an optional feature because other users want it, even if you personally find their use of it to be misguided or ill-advised?

See point 1. in short of course I just think the docs should be clear about the risks.

It seems emotion might have been a little high and am only making these points so bisync can become the best possible version and to help aid development not to dig, insult or belittle. We are all constantly learning. I do massively appreciate your work and time on this #7410 is just one example of being frickin awesome and has completely changed the viability of this tool for many.

I am away again this weekend and from Friday next week family time starts so if we don't speak before then; Merry Christmas and a Happy New Year ! 🎅 🎄 🎉

n.b. I will get to testing #7470 though it might not be until the new year now

**Comment by nielash** [comment]
No worries, and thank you as always for taking the time to discuss this all in such great detail during such a busy time! It is valuable and I really do appreciate it, even when I don't always share your view 100% of the time.

I think that there is actually a lot that we do agree on here, and if we reach a lot of the same conclusions for different reasons, that is ok 🙂 Here is what I think the main takeaways are, for me:

- You are right that the default should be changed from the legacy behavior to the auto-numbering system that you have proposed.
- The legacy behavior should remain available for the minority of users who want it, but should not be the default, much less the _only_ option.
- The legacy behavior does indeed carry some risk of having older conflicts overwritten by newer conflicts, and at minimum, this risk should be made more clear in the docs. I also remain open to other changes, if you have suggestions as to how bisync could handle this scenario more safely for the users who choose to use it.
- Bisync does not replace the need for a proper 3-2-1 incremental backup system -- they are different tools for different purposes.
- Your `-cr` and `-cs` flags are great ideas and should definitely be added -- I look forward to working on them, and probably pretty soon!

If we don't speak again before then, a Merry Christmas and Happy New Year to you as well! 🥂

**Comment by nielash** [comment]
Hi @martind69, I have [posted a preview here](https://github.com/nielash/rclone/releases/tag/v1.65.0-beta.7566.68eb6f5b2.bisync-feature-preview-2024-01-01) -- feel free to try it if you like and let me know if you have feedback. It includes my proposed implementation of the conflict-resolve features discussed on this ticket, and also [many](https://github.com/nielash/rclone/blob/bisync-feature-preview-2024-01-01/docs/content/bisync.md#v166) other [fixes](https://github.com/nielash/rclone/compare/668711e43213bc9464c0b25c88621cd987898f08...bisync-feature-preview-2024-01-01) that are not yet in the latest official rclone beta.

> Please consider to accept `--retries*` with bisync!

One of my proposed commits does exactly this 🙂 https://github.com/nielash/rclone/commit/f1109f1da86a4e43d47f18617523a4a1eed842dd
Note that it requires using `--resilient`, the default is `--retries 3`, and `--retries-sleep` is not yet implemented (but it would be easy to add this -- do you want to open a new issue for this?)

Given what you described, I think you may also be interested in the [`--recover`](https://github.com/nielash/rclone/blob/bisync-feature-preview-2024-01-01/docs/content/bisync.md#--recover) feature proposed on #7501. I would recommend trying it with something like:

```
--resilient --recover --max-lock 2m --conflict-resolve newer
```
and `--resync-mode newer` (only when resyncing.)

I would probably suggest _not_ using `--conflict-loser delete` right away, until you have used `--conflict-resolve newer` for a little while and are comfortable with how it works and you are sure that deleting the loser is what you want. I also highly suggest using [`--backup-dir1` and `--backup-dir2`](https://github.com/nielash/rclone/blob/bisync-feature-preview-2024-01-01/docs/content/bisync.md#--backup-dir1-and---backup-dir2) if you are going to use `--conflict-loser delete`.

Let me know how it works for you if you give it a try 🙂

**Comment by Goudrons** [comment]
Is this feature finally released?

**Comment by martind69** [comment]
> Is this feature finally released?

Please read before asking questions. See issue Nr. 7529

**Comment by Goudrons** [comment]
Don't forget that there are also newbies on GitHub who are not trained computer scientists. If beginners are not welcome here and are dismissed with snide remarks, then I'll be gone in a flash. Do you really think I didn't search before asking a question? You seem to assume that I did. A link to the relevant section and an explanation of where to find it would have had a completely different, less unpleasant effect.
But thanks anyway.

**Comment by nielash** [comment]
Yes, this feature was released in 2023. https://github.com/rclone/rclone/pull/7529

Here is a link to the relevant section of the docs: https://rclone.org/bisync/#conflict-resolve

**Comment by Goudrons** [comment]
Thank you nielash, i‘ll give it a try
